### PR TITLE
PRO-1017: Add Cash Web View

### PR DESCRIPTION
**Pull request by Yash-Chaubal**
https://linear.app/pillarproject/issue/PRO-1017/pillar-wallet-add-cash-web-view

## Description

Added Mt Pelerin integration to the Add Cash flow. Use deployed Matic account for the signature, auto deploys account if not already deployed.

## How Has This Been Tested?
Tested locally on an iPhone 11 simulator and a Google Pixel 3 simulator.

## Screenshots (if appropriate):

https://user-images.githubusercontent.com/55538608/223079356-87c1df89-7176-4e6a-8801-4e2915caf754.mp4

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/actions/accountsActions.js [hunk 1]
```diff
@@ -309,7 +309,7 @@ export const deployAccounts = () => {
 
     dispatch({ type: DEPLOY_ACCOUNTS_FETCHING, payload: true });
 
-    const networkPromises = networkChains.map((chain) => etherspotServices.setbatchDeployAccount(chain));
+    const networkPromises = networkChains.map((chain) => etherspotServices.setBatchDeployAccount(chain));
```
Comment (IAmKio) [inline src/actions/accountsActions.js:312]: ooofff, good spot

File: src/services/etherspot.js [hunk 2]
```diff
@@ -480,9 +480,13 @@ export class EtherspotService {
...
 
-    const { hash } = await sdk.submitGatewayBatch();
-
-    await this.waitForTransactionHashFromSubmittedBatch(chain, hash);
+    try {
+      const { hash } = await sdk.submitGatewayBatch();
+      if (returnHash) return hash;
+      await this.waitForTransactionHashFromSubmittedBatch(chain, hash);
+    } catch (e) {
+      return AccountStates.UnDeployed;
```
Comment (IAmKio) [inline src/services/etherspot.js:488]: Shall we add an error log here?

File: src/utils/fiatToCrypto.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+// Pelerin
+export const buildMtPelerinOptions = (code: string, address: string) => {
+  let onRampOptions = {
+    lang: 'en',
+    tab: 'buy',
+    tabs: 'buy',
+    chain: 'xdai_mainnet',
+    net: 'matic_mainnet',
+    nets: 'arbitrum_mainnet,avalanche_mainnet,bsc_mainnet,fantom_mainnet,mainnet,optimism_mainnet,xdai_mainnet',
+    crys: 'AVAX,BNB,BTCB,BUSD,DAI,ETH,FRAX,LUSD,MAI,MATIC,RBTC,RDOC,RIF,USDC,USDT,WBTC,WETH,XCHF,XDAI,XTZ',
```
Comment (IAmKio) [inline src/utils/fiatToCrypto.ts:98]: that's a weird property name 😂

File: src/utils/fiatToCrypto.ts [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+  if (account?.state === AccountStates.UnDeployed) {
+    if (deployingAccount) {
+      !!showAlert && showAlert('Deploying Etherspot wallet, please try again later.');
+      return;
+    }
+
+    !!setDeployingAccount && setDeployingAccount(true);
+    const hash: string = await etherspotService.setBatchDeployAccount(chain, true);
+    !!setDeployingAccount && setDeployingAccount(false);
```
Comment (IAmKio) [inline src/utils/fiatToCrypto.ts:133]: it looks like `setBatchDeployAccount` may possibly return an error if one of the `await`'s fail, however i think that isn't going to evaluate into a `false`y value. Should we check `typeof` or `try` / `catch` here?

File: src/utils/fiatToCrypto.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    if (!hash || hash === AccountStates.Deployed || hash === AccountStates.UnDeployed) {
+      !!showAlert && showAlert('Failed to deploy the Etherspot wallet');
```
Comment (IAmKio) [inline src/utils/fiatToCrypto.ts:136]: Please add a suggestion what to do next after the message, such as try again later. This string should also be in the i18n files.

**Review by IAmKio — CHANGES_REQUESTED** [review]
Just a few small points, but other than that - great work all round!

File: src/configs/envConfig.js [hunk 6]
```diff
@@ -88,6 +89,7 @@ const envVars = {
     COLLECTIBLES_NETWORK: 'rinkeby',
     OPEN_SEA_API: 'https://rinkeby-api.opensea.io/api/v1',
     RAMPNETWORK_WIDGET_URL: 'https://ri-widget-staging-kovan.firebaseapp.com/',
+    PELERIN_WIDGET_URL: 'https://buy.mtpelerin.com/',
```
[the remark this answers is not in the document]
Comment (Yash-Chaubal) [reply src/configs/envConfig.js:92]: I didn't really see anything in the docs (https://developers.mtpelerin.com/integration-guides/options), but I dunno if that's located somewhere else

File: src/utils/fiatToCrypto.ts (hunk 4 above)
Comment (Yash-Chaubal) [reply src/utils/fiatToCrypto.ts:133]: Good catch, yeah I'll see if the errors we return have any identifying parts, or I'll just return null if it's an error

File: src/utils/fiatToCrypto.ts (hunk 3 above)
Comment (Yash-Chaubal) [reply src/utils/fiatToCrypto.ts:98]: lmao i've seen assets, tokens, even coins - but never crys, maybe the 2 saved chars make a difference

File: src/screens/AddCash/modal/AddCashModal.tsx [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+  useEffect(() => {
+    return () => {
+      setDeployingAccount(false);
+      setDeployError('');
+    };
+  }, []);
+
+  const openPelerin = async () => {
+    const url = await pelerinWidgetUrl(deployingAccount, setDeployingAccount, setDeployError);
+    setAddCashUrl?.(url);
```
[the remark this answers is not in the document]
Comment (Yash-Chaubal) [reply src/screens/AddCash/modal/AddCashModal.tsx:63]: We already show users an alert I believe, but I'll add some error logs when we display the alert

File: src/utils/fiatToCrypto.ts (hunk 4 above)
Comment (Yash-Chaubal) [reply src/utils/fiatToCrypto.ts:133]: Just checked out the `setBatchDeployAccount` function - we never return an error, just log them. The only thing that's ever returned is null, a hash, or `AccountStates`, which we then handle in the later if statement

File: src/configs/envConfig.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/configs/envConfig.js:92]: No worries, just curious. I didn't immediately see anything either.

File: src/screens/AddCash/modal/AddCashModal.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/screens/AddCash/modal/AddCashModal.tsx]: Thanks

File: src/utils/fiatToCrypto.ts (hunk 4 above)
Comment (IAmKio) [reply src/utils/fiatToCrypto.ts:134]: Gotcha, thanks for checking - in that case we're all good!

**Review by IAmKio — APPROVED** [review]
LGTM!

**Pull request merged**